config/package_resouce.go

```diff
@@ -14,26 +14,229 @@
+type ManagedPackage struct {
+	Apt    AptPackage
+	Deb    DebPackage
+	GooGet GooGetPackage
+	MSI    MSIPackage
+	Yum    YumPackage
+	Zypper ZypperPackage
+	RPM    RPMPackage
 }
 
 func (p *packageResouce) validate() (*ManagedResources, error) {
-	// TODO: implement
-	return nil, nil
+	switch p.GetSystemPackage().(type) {
+	case *agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_Apt:
+		pr := p.GetApt()
+		if !packages.AptExists {
+			return nil, fmt.Errorf("cannot manage Apt package %q because apt-get does not exist on the system", pr.GetName())
+		}
+
+		switch p.GetDesiredState() {
+		case agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_INSTALLED:
+			p.policy.Apt.Install = pr
+		case agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_REMOVED:
+			p.policy.Apt.Remove = pr
+		}
+	case *agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_Deb_:
+		pr := p.GetDeb()
+		if !packages.DpkgExists {
+			return nil, fmt.Errorf("cannot manage Deb package because dpkg does not exist on the system")
+		}
+
+		switch p.GetDesiredState() {
+		case agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_INSTALLED:
+			p.policy.Deb.Install = pr
+		case agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_REMOVED:
+			p.policy.Deb.Remove = pr
+		}
+	case *agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_Msi:
+		pr := p.GetMsi()
+		if !packages.MSIExecExists {
+			return nil, fmt.Errorf("cannot manage MSI package because msiexec does not exist on the system")
+		}
+
+		switch p.GetDesiredState() {
+		case agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_INSTALLED:
+			p.policy.MSI.Install = pr
+		case agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_REMOVED:
+			p.policy.MSI.Remove = pr
+		}
+	case *agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_Googet:
+		pr := p.GetGooget()
+		if !packages.GooGetExists {
+			return nil, fmt.Errorf("cannot manage GooGet package %q because googet does not exist on the system", pr.GetName())
+		}
+
+		switch p.GetDesiredState() {
+		case agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_INSTALLED:
+			p.policy.GooGet.Install = pr
+		case agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_REMOVED:
+			p.policy.GooGet.Remove = pr
+		}
+	case *agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_Yum:
+		pr := p.GetYum()
+		if !packages.YumExists {
+			return nil, fmt.Errorf("cannot manage Yum package %q because yum does not exist on the system", pr.GetName())
+		}
+
+		switch p.GetDesiredState() {
+		case agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_INSTALLED:
+			p.policy.Yum.Install = pr
+		case agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_REMOVED:
+			p.policy.Yum.Remove = pr
+		}
+	case *agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_Zypper_:
```

Comment:
nit: Is the underscore at the end expected? Not seeing for other package managers.

Reply:
It's an artifact of the autogenerated client library, has to do with the wrapped type and lowercase vs uppercase